aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -542,6 +542,97 @@ Tensor &nuclear_norm_out(Tensor& result, const Tensor& self, bool keepdim) {
   return at::sum_out(result, std::get<1>(at::svd(self)), 0, keepdim);
 }
 
+// Return a permutation with the given axes moved to the end.
+static Tensor move_to_end(const Tensor& self, IntArrayRef axes) {
+  const std::vector<int64_t> a = axes.vec();
+  std::vector<int64_t> perm;
+
+  for (int64_t i = 0; i < self.ndimension(); i++) {
+    auto it = std::find(a.begin(), a.end(), i);
+    if (it == a.end()) {
+       perm.push_back(i);
+    }
+  }
+  for (auto i : a) {
+    perm.push_back(i);
+  }
+
+  return self.permute(perm);
+}
+
+// Non-optimized batched svd implementation. This can be merged with at::svd
+// once at::svd has been ported to ATen.
+static std::tuple<Tensor, Tensor, Tensor>
+batch_svd(const Tensor& self, bool some, bool compute_uv)
+{
+  const int64_t ndim = self.ndimension();
+
+  TORCH_CHECK(
+      ndim >= 2,
+      "Expected a tensor with at least 2 dimensions, but got a ",
+      ndim,
+      " dimensions tensor instead.");
+
+  if (ndim == 2) {
+    return at::svd(self, some, compute_uv);
+  }
+
+  const int64_t n = self.size(-2);
+  const int64_t m = self.size(-1);
+  const int64_t k = std::min<int64_t>(n, m);
+  const int64_t nn = (some && compute_uv) ? k : n;
+  const int64_t mm = (some && compute_uv) ? k : m;
+  int64_t p = 1;
+
+  for (int64_t i = 0; i < ndim-2; i++) {
+    p *= self.size(i);
+  }
+
+  Tensor t = self.reshape({p, n, m}).contiguous();
```

Comment:
I think`reshape` makes `contiguous` call redundant.

Reply:
If it's the same as on the Python level, I think the result of reshape can still be non-contiguous:

```
>>> x = torch.arange(20)
>>> y = x[::2]
>>> z = y.reshape(2,5)
>>> z.is_contiguous()
False
```

Reply:
On the other hand, if `at::svd()` can handle non-contiguous matrices and `t[i]` is well defined for non-contiguous outer dimensions, it should work.  I wasn't sure about `t[i]`.

Reply:
Oh I’m sorry. I was under the view that reshape returns contiguous tensors.

Reply:
at::svd can handle non-contiguous matrices, since a column major clone is performed internally to satisfy LAPACK input requirements. t[i] is effectively a slice op which also should be well defined for non-contiguous tensors (but I am not sure).